lib/proto/proto.go

```diff
@@ -0,0 +1,1303 @@
+		err = proto.Unmarshal(data, m.Message())
+	} else {
+		err = prototext.Unmarshal(data, m.Message())
+	}
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalling %s failed: %v", desc.FullName(), err)
+	}
+	return m, nil
+}
+
+func (m *Message) String() string {
+	buf := new(bytes.Buffer)
+	buf.WriteString(string(m.desc().FullName()))
+	buf.WriteByte('(')
+
+	// Sort fields (including extensions) by number.
+	var fields []protoreflect.FieldDescriptor
+	m.msg.Range(func(fdesc protoreflect.FieldDescriptor, v protoreflect.Value) bool {
+		// TODO(adonovan): opt: save v in table too.
+		fields = append(fields, fdesc)
+		return true
+	})
+	sort.Slice(fields, func(i, j int) bool {
+		return fields[i].Number() < fields[j].Number()
+	})
+
+	for i, fdesc := range fields {
+		if i > 0 {
+			buf.WriteString(", ")
+		}
+		if fdesc.IsExtension() {
+			// extension field: "[pkg.Msg.field]"
+			buf.WriteString(string(fdesc.FullName()))
+		} else if fdesc.Kind() != protoreflect.GroupKind {
+			// ordinary field: "field"
+			buf.WriteString(string(fdesc.Name()))
+		} else {
+			// group field: "MyGroup"
+			//
+			// The name of a group is the mangled version,
+			// while the true name of a group is the message itself.
+			// For example, for a group called "MyGroup",
+			// the inlined message will be called "MyGroup",
+			// but the field will be named "mygroup".
+			// This rule complicates name logic everywhere.
+			buf.WriteString(string(fdesc.Message().Name()))
+		}
+		buf.WriteString("=")
+		writeString(buf, fdesc, m.msg.Get(fdesc))
+	}
+	buf.WriteByte(')')
+	return buf.String()
+}
+
+func (m *Message) Type() string                { return "proto.Message" }
+func (m *Message) Truth() starlark.Bool        { return true }
+func (m *Message) Freeze()                     { *m.frozen = true }
+func (m *Message) Hash() (h uint32, err error) { return uint32(uintptr(unsafe.Pointer(m))), nil } // identity hash
+
+// Attr returns the value of this message's field of the specified name.
+// Extension fields are not accessible this way as their names are not unique.
+func (m *Message) Attr(name string) (starlark.Value, error) {
+	// The name 'descriptor' is already effectively reserved
+	// by the Go API for generated message types.
+	if name == "descriptor" {
+		return MessageDescriptor{Desc: m.desc()}, nil
+	}
+
+	fdesc, err := fieldDesc(m.desc(), name)
+	if err != nil {
+		return nil, err
+	}
+	return m.getField(fdesc), nil
+}
+
+func (m *Message) getField(fdesc protoreflect.FieldDescriptor) starlark.Value {
+	if fdesc.IsExtension() {
+		// See explanation in setField.
+		fdesc = dynamicpb.NewExtensionType(fdesc).TypeDescriptor()
+	}
+
+	if m.msg.Has(fdesc) {
+		return toStarlark(fdesc, m.msg.Get(fdesc), m.frozen)
+	}
+	return defaultValue(fdesc)
+}
+
+//go:linkname detrandDisable google.golang.org/protobuf/internal/detrand.Disable
+func detrandDisable()
+
+func init() {
+	// Nasty hack to disable the randomization of output that occurs in textproto.
```

Comment:
Yes, it is brittle---but I hope that it should break audibly, so to speak. If detrand goes away, this won't link, and our tests will fail. At that point, if the proposal has been submitted, then we will use the new standard format, and if it has been rejected, I will have to write a new text encoder.

Reply:
I think `linkname` should work with modules, though I'm not completely sure. It may even be more predictable than GOPATH mode, since we won't have vendor directories changing the linked package path.